classes/privacy/provider.php

```diff
@@ -0,0 +1,225 @@
+                continue;
+            }
+            $sql = "SELECT *
+                  FROM {local_lsf_course}
+                 WHERE (requesterid = :userid)";
+            $contextparams['userid'] = $user->id;
+
+            $coursesrequested = $DB->get_recordset_sql($sql, $contextparams);
+            $sql = "SELECT *
+                  FROM {local_lsf_course}
+                 WHERE (acceptorid = :userid)";
+            $coursesaccepted = $DB->get_recordset_sql($sql, $contextparams);
+
+            if (empty($courserequested) && empty($coursesaccepted)) {
+                return;
+            }
+
+            foreach ($coursesrequested as $courserequest) {
+                $contextdatatowrite[] = self::create_data_to_write('requester', $courserequest, $user->username);
+            }
+            foreach ($coursesaccepted as $courserequest) {
+                $contextdatatowrite[] = self::create_data_to_write('acceptor', $courserequest, $user->username);
+            }
+        }
+        $subcontext = [
+            get_string('pluginname', 'local_lsf_unification')
+        ];
+        if (!empty($contextdatatowrite)) {
+            writer::with_context($context)->export_data($subcontext, (object)$contextdatatowrite);
+        }
+    }
+
+
+    /**
+     * Helper Function to build the data for a given courserequest.
+     * @param $action \string requester or acceptor
+     * @param $courserequest
+     * @param $username
+     * @return object
+     * @throws \coding_exception
+     * @throws \dml_exception
+     */
+    private function create_data_to_write($action, $courserequest, $username) {
+        global $DB;
+        $status = '';
+        switch ($courserequest->requeststate) {
+            case 0 :
+                $status = "is declined or not requested";
+                break;
+            case 1 :
+                $status =
+                    "is waiting";
+                break;
+            case 2 :
+                $status =
+                    "is granted";
+                break;
+        }
+        $data = (object)[
+            'action' => get_string('privacy:local_lsf_unification:' . $action, 'local_lsf_unification'),
+            'timestamp' => date('m/d/Y H:i:s', $courserequest->timestamp),
+            'username' => $username,
+            'status' => $status
+        ];
+        // In case the course exist we add the name to the information.
+        $course = $DB->get_record('course', array('id' => $courserequest->mdlid));
+        if (!empty($course)) {
+            $data->coursename = $course->fullname;
+        }
+        return $data;
+    }
+
+    /**
+     * Deletes data from the table for all users in all contexts.
+     * @param $context
+     * @throws \dml_exception
+     */
+    public static function _delete_data_for_all_users_in_context($context) {
+        global $DB;
+
+        // Sanity check that context is at the System context level.
```

Comment:
Comment is not accurate anymore.

In addition, also needs code for the case that `$context` is the system context; i.e. purge everything.